rllib/env/remote_base_env.py

```diff
@@ -0,0 +1,225 @@
+            if len(self.existing_envs) > 0 and isinstance(
+                    self.existing_envs[0], ray.actor.ActorHandle):
+                self.make_env_creates_actors = True
+                self.actors = []
+                while len(self.actors) < self.num_envs:
+                    self.actors.append(self.make_env(len(self.actors)))
+            # `self.make_env` produces gym.Envs (or children thereof, such
+            # as MultiAgentEnv): Need to auto-wrap it here. The problem with
+            # this is that custom methods wil get lost. If you would like to
+            # keep your custom methods in your envs, you should provide the
+            # env class directly in your config (w/o tune.register_env()),
+            # such that your class will directly be made a @ray.remote
+            # (w/o the wrapping via `_Remote[Multi|Single]AgentEnv`).
+            else:
+
+                def make_remote_env(i):
+                    logger.info("Launching env {} in remote actor".format(i))
+                    if self.multiagent:
+                        return _RemoteMultiAgentEnv.remote(self.make_env, i)
+                    else:
+                        return _RemoteSingleAgentEnv.remote(self.make_env, i)
+
+                self.actors = [
+                    make_remote_env(i) for i in range(self.num_envs)
+                ]
+
+        if self.pending is None:
+            self.pending = {a.reset.remote(): a for a in self.actors}
+
+        # each keyed by env_id in [0, num_remote_envs)
+        obs, rewards, dones, infos = {}, {}, {}, {}
+        ready = []
+
+        # Wait for at least 1 env to be ready here
+        while not ready:
+            ready, _ = ray.wait(
+                list(self.pending),
+                num_returns=len(self.pending),
+                timeout=self.poll_timeout)
+
+        # Get and return observations for each of the ready envs
+        env_ids = set()
+        for obj_ref in ready:
+            actor = self.pending.pop(obj_ref)
+            env_id = self.actors.index(actor)
+            env_ids.add(env_id)
+            ret = ray.get(obj_ref)
+            # Our sub-envs are simple Actor-turned gym.Envs or MultiAgentEnvs.
+            if self.make_env_creates_actors:
+                rew, done, info = None, None, None
+                if self.multiagent:
+                    if isinstance(ret, tuple) and len(ret) == 4:
+                        ob, rew, done, info = ret
+                    else:
+                        ob = ret
+                else:
+                    if isinstance(ret, tuple) and len(ret) == 4:
+                        ob = {_DUMMY_AGENT_ID: ret[0]}
+                        rew = {_DUMMY_AGENT_ID: ret[1]}
+                        done = {_DUMMY_AGENT_ID: ret[2], "__all__": ret[2]}
+                        info = {_DUMMY_AGENT_ID: ret[3]}
+                    else:
+                        ob = {_DUMMY_AGENT_ID: ret}
+
+                if rew is None:
```

Comment:
is this a fix for a specific env/version or a general safety net for user's custom envs?
feel like this may be concealing bugs if they forgot to return reward, for example.

Reply:
Great question. It's when we have a return value from `reset()` (rather than `step()`). Reset does not return rewards (only 1 return value; step=4 return values).
Added a comment explaining this.